---
title: "Differentially expressed genes for naphthenic acid degradation"
author: "Nweze Julius"
output: html_document
date: "2024-09-17"
output:
  rmarkdown::html_document:
    code_folding: show
    dev: png
    df_print: kable
    fig_caption: yes
    highlight: pygments
    keep_md: yes
    number_sections: no
    theme: flatly
    toc: yes
    toc_depth: 5
    toc_float: yes
  html_document:
    df_print: paged
    toc: yes
    toc_depth: '5'
editor_options: 
  chunk_output_type: console
---

```{r libraries, include=F}
repo <- "http://cran.wu.ac.at"
lib.loc <- Sys.getenv("R_LIBS_USER")

update.packages(
    lib.loc, 
    repos = repo,
    ask = FALSE
)


.cran_libs <- c(
"devtools", # The tools required to efficiently manage other packages
"data.table", # Optimized for fast reading of large files (fread)
"readr", # For reading .csv files efficiently
"tibble", # For tibble manipulations (data frames with extra features)
"readODS", # Read ODS Files
"kableExtra",
"tidyverse", # Helps to transform and better present data. 
"ggplot2", # For creating declarative and customizable graphics
"readxl",
"magrittr", # Forward and backward “pipe”-like operator
"dplyr", # A grammar of data manipulation
"DESeq2", # For RNA-seq count data analysis and differential expression
"ggnewscale") # To allow multiple fill scales

.inst <- .cran_libs %in% installed.packages()
if (any(!.inst)) {
   install.packages(.cran_libs[!.inst],
                    repos = repo,
                    lib = lib.loc)
}


.bioc_libs <- c(
  #"multtest", #Resampling-based multiple hypothesis testing
)

.bioc_inst <- .bioc_libs %in% installed.packages()
if (any(!.bioc_inst)) {
   if (!requireNamespace("BiocManager", quietly = TRUE))
   install.packages("BiocManager")
   BiocManager::install(ask = F, lib = lib.loc)  # upgrade bioC packages
   BiocManager::install(.bioc_libs[!.bioc_inst], ask = F, lib = lib.loc)
}

.local_libs <- c()

.inst <- names(.local_libs) %in% installed.packages()
if (any(!.inst)) {
   install.packages(paste0("~/R/", .local_libs[!.inst]) ,repos = NULL, type = "source", lib = lib.loc)
}

.github_libs <- c(
  "wilkelab/ggtext", # Improved text rendering support for 'ggplot2' 
  "ACCLAB/dabestr" # Data Analysis using Bootstrap-Coupled Estimation 
)

.github_lib_names <- stringr::str_replace(.github_libs, ".*/(.*)$", "\\1")

.github_inst <- .github_lib_names %in% installed.packages()
if (any(!.github_inst)) {
  devtools::install_github(.github_libs[!.github_inst],
                           lib = lib.loc,
                           dependencies = TRUE)
}

# Load packages into session, and print package version
(loaded.libs <- sapply(c(.cran_libs, .bioc_libs, names(.local_libs), .github_lib_names), require, character.only = TRUE))
if (!all(loaded.libs)) {stop(paste("Package(s):", names(loaded.libs[loaded.libs == FALSE]), "could not be loaded"))}
sapply(c(.cran_libs, .bioc_libs, names(.local_libs), .github_lib_names), packageVersion)
```


**Load metadata**
```{r load data, cache = T}
# Load the metadata
metadata <- read_tsv("metadata.tsv")
metadata_full <- read_xlsx("metadata_Root.xlsx")

# Merge metadata
metadata <-  metadata %>% dplyr::left_join(metadata_full, by=c("SampleID", "Time"), multiple = "all", relationship = "many-to-many")

# Save merged metadata
# write.csv(metadata, "Complete_metadata.csv")

# load the abundance in cpm
data <- read_tsv("merged_gene_abundance_cpm.tsv", name_repair = "minimal")

# Reshape abundance from wide to long
long_data <- pivot_longer(
  data,
  cols = -1,                # Keep the first column (likely gene names) as is
  names_to = "Sample",       # Name for sample column
  values_to = "CPM",
  names_repair = "minimal"   # Name for CPM values
)

# Save the MCycDB_abundance
# write.csv(long_data, "long_data_cpm.csv")

# Merge metadata and abundance
MCycDB_abundance <-  metadata %>% dplyr::left_join(long_data, by=c("Sample"), multiple = "all", relationship = "many-to-many")

# Save the MCycDB_abundance
write.csv(MCycDB_abundance, "MCycDB_abundance.csv")

# Load annotation
read_ods("annotations.ods", sheet = "annotations",  col_names = TRUE) ->
Taxa


# Merge metadata, abundance and annotation
MCycDB_abundance_tax <-  MCycDB_abundance %>% dplyr::left_join(Taxa, by=c("Gene_ID"), multiple = "all", relationship = "many-to-many")


# Save the MCycDB_abundance
write.csv(MCycDB_abundance_tax, "MCycDB_abundance_tax.csv")


# Load the gene
gene <- read_tsv("MCycDB.gene.tsv")

# Load the gene
Path <- read_tsv("MCycDB.pathway.tsv")

#read_ods("MCycDB.filtered.ods", sheet = "Pathway",  col_names = TRUE) ->
#Path

# Load MCycDB_genes_converted
MCycDB_genes <- read_tsv("MCycDB_genes.tsv")
#read_ods("MCycDB.filtered.ods", sheet = "MCycDB_genes",  col_names = TRUE) ->
#MCycDB_gene

# Combine the gene code and the pathway
Path_genes <-  Path %>% dplyr::left_join(gene, by=c("Gene"), multiple = "all", relationship = "many-to-many") %>% drop_na(Gene)

# Save the Path_genes
write.csv(Path_genes, "Path_genes.csv")

# Combine the identified genes, gene code and the pathway
MCycDB_gene_path <-  MCycDB_genes %>% dplyr::left_join(Path_genes, by=c("qseqid"), multiple = "all", relationship = "many-to-many") %>% drop_na(Gene)

# Save the MCycDB_gene_path
write.csv(MCycDB_gene_path, "MCycDB_gene_path.csv")

# Merge metadata, abundance and annotation
MCycDB_abundance_tax_blast <-  MCycDB_abundance_tax %>% dplyr::left_join(MCycDB_gene_path, by=c("Gene_ID"), multiple = "all", relationship = "many-to-many") %>% drop_na(Pathway) %>% subset(tax_kingdom %in% c("Archaea", "Bacteria", "Algae", "Fungi"))

# Save the MCycDB_abundance
write.csv(MCycDB_abundance_tax_blast, "MCycDB_abundance_tax_blast.csv")


# Save an object to a file
saveRDS(MCycDB_abundance_tax_blast, file = "MCycDB_abundance_tax_blast.rds")
# Restore the object
MCycDB_abundance_tax_blast <- readRDS(file = "MCycDB_abundance_tax_blast.rds")
```


**Plot abundance of NA-degrading genes in OSPW and RO**
```{r bar plot data, cache = T}

path_colour <- c("Central methanogenic pathway" = "#0000FF", "Aceticlastic methanogenesis" = "#A52A2A", "Hydrogenotrophic methanogenesis" = "#DEB887", "Methylotrophic methanogenesis" = "#838B8B", "Anaerobic oxidation of methane (AOM)" = "#7FFF00", "Oxidation of merthane and C1 compounds" = "#D2691E", "Oxidation of formaldehyde" = "#8B3E2F", "Serine cycle" = "#FF8C00", "RuMP cycle" = "#FF1493", "Oxidation of formate" = "#FFD700")

Ab_NA <- MCycDB_abundance_tax_blast %>%
  # You can uncomment the filter if you want OSPW only
  mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  
    # Step 1: Average again per Gene_codes
  group_by(Water_type, Time, Gene) %>%
  summarise(cpm = mean(CPM), .groups = "drop")

Ab_NA <-  Ab_NA %>% dplyr::left_join((unique(Path_genes %>% dplyr::select(Gene, Pathway))), by=c("Gene"), multiple = "all", relationship = "many-to-many") %>% drop_na(Water_type)

Ab_NA %>%
mutate(Pathways = factor(Pathway, levels = c("Central methanogenic pathway", "Aceticlastic methanogenesis", "Hydrogenotrophic methanogenesis", "Methylotrophic methanogenesis", "Anaerobic oxidation of methane (AOM)", "Oxidation of merthane and C1 compounds", "Oxidation of formaldehyde", "Serine cycle", "RuMP cycle", "Oxidation of formate"))) %>%
mutate(Gene = factor(Gene, levels = c("mtrA1","mtrB1","mtrC1","mtrD1","mtrE1","mtrF1","mtrG1","mtrH1","mcrA1","mcrB1","mcrC1","mcrD1","mcrG1","HdrA1_1","hdrB1_1","HdrC1_1","HdrA2_1","HdrB2_1","HdrC2_1","HdrD_1","HdrE_1","mvhA","mvhD","mvhG","ehaA","ehaB","ehaC","ehaD","ehaE","ehaF","ehaG","ehaH","ehaI","ehaJ","ehaK","ehaL","ehaM","ehaN","ehaO","ehaP","ehaQ","ehaR","ehbA","ehbB","ehbC","ehbD","ehbE","ehbF","ehbG","ehbH","ehbI","ehbJ","ehbK","ehbL","ehbM","ehbN","ehbO","ehbP","ehbQ","mbhL","mbhK","mbhJ","rnfA1","rnfB1","rnfC1","rnfD1","rnfE1","rnfG1","echA","echB","echC","echD","echE","echF","vhoA","vhoC","vhoG","vhtA","vhtC","vhtG","vhuA","vhuU","vhuD","vhuG","vhcA","vhcD","vhcG","fpoA","fpoB","fpoC","fpoD","fpoF","fpoH","fpoI","fpoJ","fpoK","fpoL","fpoM","fpoN","fpoO","fqoA","fqoD","fqoF","fqoH","fqoJ","fqoK","fqoL","fqoM","fqoN","frhA1","frhB1","frhD1","frhG1","fruA","fruB","fruD","fruG","frcA1","frcB1","frcD1","frcG1","fmdA2","fmdB2","fmdC1","fmdD2","fmdE1","fmdF1","fwdA2","fwdB2","fwdC2","fwdD2","fwdE2","fwdF2","fwdG2","fwdH2","ftr1","mch1","mtdA2","mtdB2","hmd2","mer2","metF2","acs","acsA","acsC","acsD","ackA","pta","cdhC","cdhD","cdhE","cdhA","cdhB","acdA","acdB","cooS","cooF","mtaA","mtaB","mtaC","mtbA","mtbB","mtbC","mtsA","mtsB","torA","torC","torD","torY","torZ","mttB","mttC","mtmB","mtmC","fmdA1","fmdB1","fmdC2","fmdD1","fmdE2","fmdF2","fwdA1","fwdB1","fwdC1","fwdD1","fwdE1","fwdF1","fwdG1","fwdH1","ftr2","mch2","mtdA1","mtdB1","hmd1","mer1","metF1","mtrA2","mtrB2","mtrC2","mtrD2","mtrE2","mtrF2","mtrG2","mtrH2","mcrA2","mcrB2","mcrC2","mcrD2","mcrG2","hdrA1","hdrB1","hdrC1","hdrA2","hdrB2","hdrC2","hdrD","hdrE","rnfA","rnfB","rnfC","rnfD","rnfE","rnfG","frhA","frhB","frhD","frhG","fruA1","fruB1","fruD1","fruG1","frcA","frcB","frcD","frcG","fpoA1","fpoB1","fpoC1","fpoD1","fpoF1","fpoH1","fpoI1","fpoJ1","fpoK1","fpoL1","fpoM1","fpoN1","fpoO1","fqoA1","fqoD1","fqoF1","fqoH1","fqoJ1","fqoK1","fqoL1","fqoM1","fqoN1","narG","narZ","narH","narY","narB","napH","nrfH","nrfA","nxrA","nod","nirK","nirS","cytC","pqqA","pqqB","pqqC","pqqD","pqqE","pqqF","mmoX","mmoY","mmoZ","mmoB","mmoC","mmoD","pmoA","pmoB","pmoC","amoA","amoB","amoC","xoxF1","xoxF2","xoxF4","xoxF5","mxaF","mxaI","mxaJ","mxaG","mxaA","mxaC","mxaK","mxaL","mxaD","mauA","mauB","mauC","mauD","mauE","mauF","mgsA","mgsB","mgsC","mgdA","mgdB","mgdD","qhpA","dcmR","dcmA","mdh","tmm","gfa","frmA","frmB","fghA","fdhA-K00148","fae","mtdB3","mch3","mdo","fdm","fdwA","fdwB","fdwE","fdsB","fdsD","fdsG","fdoG","fdoH","fdoI","fdhF","fdhA","fdhB","fchA","mtdA3","dfrA1","dfrA12","dfrA10","dfrA19","folA","glyA","hprA","gck","eno","ppc","mtkA","mtkB","mcl","mdh-K00024","gpmI","gpmB","apgM","serA","serC","serB","thrH","psp","porA","porB","porD","porG","pps","fae-hps","hxlB","hxlA","pfkA","pfkB","pfkC","pfp","fbaA","fbaB","glpX","glpX.SEBP","fbp","fbp3","fbp-SEBP"))) %>%
  mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
  mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  ggplot() +
  theme_bw() + 
  geom_linerange(aes(x = cpm, y = Gene, xmin = 0, xmax = cpm), colour = "black", position = position_dodge(width = 1)) +
  geom_point(aes(x = cpm, y = Gene,  colour = Pathway), size = 12, position = position_dodge(width = 1)) +
  ylab("Genes") +  
  xlab("Relative abundance (count per million)") +
  theme(
    axis.text = element_text(size = 30, colour = "black"), axis.text.x = element_text(size = 30, colour = "black", angle = 90, hjust = 0.5, vjust = 0.5), 
    axis.ticks = element_line(size = 4), 
    axis.title = element_text(size = 35),  
    legend.title = element_text(size = 28), 
    strip.text.x = element_text(size = 26), 
    legend.text = element_text(colour = "black", size = 24)) +
   facet_wrap(~ Water_type + Time, ncol = 10) +
  xlim(0, 72) +
  scale_color_manual(values = path_colour)
  
 
ggsave("Figures/Genes_MCycDB_OSPW_RO.svg", width = 65, height = 65, units = "cm")


Ab_NA %>%
mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
group_by(Water_type, Time) %>%
arrange(.by_group = TRUE, desc(cpm)) %>%
kable(., digits = c(0, 1, 1, 1, 0, 1)) %>%
kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)



time_colour <- c("D-0" = "#8B8878", "D6" = "#458B00", "D19" = "#8B7500", "D40" = "#8B4513", "D60" = "#8B0A50")

# For gene units
Ab_NA <- MCycDB_abundance_tax_blast %>%
  # You can uncomment the filter if you want OSPW only
  mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  
    # Step 1: Average again per Gene_codes
  group_by(Water_type, Time, Units) %>%
  summarise(cpm = mean(CPM), .groups = "drop")

Ab_NA <-  Ab_NA %>% dplyr::left_join((unique(Path_genes %>% dplyr::select(Units, Pathway))), by=c("Units"), multiple = "all", relationship = "many-to-many") %>% drop_na(Water_type)

Ab_NA %>%
  mutate(Units = factor(Units, levels = c("mtrABCDEFGH","mcrABCDG","hdrA1B1C1A2B2C2DE","mvhADG","ehaABCDEFGHIJKLMNOPQR","ehbABCDEFGHIJKLMNOPQ","mbhLKJ","rnfABCDEG","echABCDEF","vhoACG","vhuAUDG","vhcADG","fpoABCDFHIJKLMNO","fqoADFHJKLMN","frhABDG","fruABDG","frcABDG","fmdABCDEF","fwdABCDEFGH","ftr","mch","mtd(AB)","hmd","mer","metF","acs","acsA","acsCD","ackA","pta","cdhCDE","cdhAB","acdAB","cooSF","mtaABC","mtbABC","mtsAB","torACDYZ","mttBC","mtmBC","narGZHYB","napH","nrfH","nrfA","nxrA","nod","nirKS","cytC","pqqABCDEF","mmoXYZBCD","pmoABC","amoABC","xoxF1F2F4F5","mxaFIJGACKLD","mauABCDEF","mgsABC","mgdABD","qhpA","dcmR","dcmA","mdh","tmm","gfa","frmAB","fghA","fdhA-K00148","fae","mtdB","mdo","fdm","fdwABE","fdsBDG","fdoGHI","fdhF","fdhAB","fchA","mtdA","dfrA1A12A10A19","folA","glyA","hprA","gck","eno","ppc","mtkAB","mcl","mdh-K00024","gpmIB","apgM","serA","serC","serB","thrH","psp","porABDG","pps","fae-hps","hxlB","hxlA","pfkABC","pfp","fbaAB","glpX","glpX-SEBP","fbp","fbp3","fbp-SEBP"
))) %>%
  mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
  mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  ggplot() +
  theme_bw() + 
  geom_linerange(aes(x = cpm, y = Units, xmin = 0, xmax = cpm), colour = "black", position = position_dodge(width = 1)) +
  geom_point(aes(x = cpm, y = Units,  colour = Time), size = 12, position = position_dodge(width = 1)) +
  ylab("Gene Units") +  
  xlab("Relative abundance (count per million)") +
  theme(
    axis.text = element_text(size = 30, colour = "black"), axis.text.x = element_text(size = 30, colour = "black", angle = 90, hjust = 0.5, vjust = 0.5), 
    axis.ticks = element_line(size = 4), 
    axis.title = element_text(size = 35),  
    legend.title = element_text(size = 28), 
    strip.text.x = element_text(size = 26), 
    legend.text = element_text(colour = "black", size = 24)) +
   facet_wrap(~ Water_type + Time, ncol = 10) +
  xlim(0, 72) + 
  scale_colour_manual(values = time_colour)
 
 
ggsave("Figures/Genes_MCycDB_OSPW_RO_gene_units.svg", width = 65, height = 55, units = "cm")


Ab_NA %>%
mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
group_by(Water_type, Time) %>%
arrange(.by_group = TRUE, desc(cpm)) %>%
kable(., digits = c(0, 1, 1, 1, 0, 1)) %>%
kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)

```



**Plot phylum-level taxonomic origins of genes from OSPW and RO**
```{r Taxa origins, cache = T}
# At phylum level
# Define custom colors for each Phylum
phylum_colors <- c(
  "Bacteroidota" = "#9467bd",
  "Pseudomonadota" = "#8B0A50",        # Light pink
  "Acidobacteriota" = "#1f77b4",       # Blue
  "Unclassified bacteria" = "#aec7e8", # Light blue-gray
  "Verrucomicrobiota" = "#dbdb8d",     # Light yellow
  "Cyanobacteriota" = "#8c564b",       # Gray
  "Actinomycetota" = "#ff7f0e",        # Orange
  "Euryarchaeota" = "#bcbd22",         # Yellow-green
  "Armatimonadota" = "#2ca02c",        # Green
  "Gemmatimonadota" = "#17becf",       # Light blue
  "Bacillota" = "#d62728",             # Red
  "Planctomycetota" = "#c49c94",       # Light brown
  "Aquificae" = "#c5b0d5",             # Pale purple
  "Chloroflexota" = "#e377c2",         # Pink
  "Nitrospirota" = "#A52A2A",          # Dark red
  "Spirochaetota" = "#98df8a",         # Pale green
  "Candidatus Saccharibacteria" = "#c7c7c7",  # Brown
  "Thermotogae" = "#7FFF00",           # Bright green
  "Candidatus Parcubacteria" = "#9edae5",      # Pale blue
  "candidate division WWE3" = "#0000FF",       # Pure blue
  "Candidatus Bipolaricaulota" = "#ff9896",    # Pale red
  "Candidatus Rokubacteria" = "#ffbb78",       # Pale orange
  "Unclassified archaea" = "#CDC9A5",          # Beige
  "Others (< 1%)"  = "#030303",                # Black
  "Unclassified viruses" = "#8b008b",          # Dark magenta
  "Unclassified organisms" = "#dda0dd",        # Plum
  "Endomyxa" = "#483d8b",                      # Dark slate blue
  "Bigyra" = "#ff6347",                        # Tomato red
  "Evosea" = "#40e0d0",                        # Turquoise
  "Choanoflagellata" = "#4682b4",              # Steel blue
  "Nematoda" = "#8b4513",                      # Saddle brown
  "Arthropoda" = "#ff4500",                    # Orange red
  "Chlorophyta" = "#32cd32",                   # Lime green
  "Xanthophyceae" = "#ffd700",                 # Gold
  "Bacillariophyta" = "#5f9ea0",               # Cadet blue
  "Ascomycota" = "#ffb6c1",                    # Light pink
  "Oomycota" = "#a0522d",                      # Sienna
  "Mucoromycota" = "#d2691e",                  # Chocolate
  "Unclassified fungi" = "#8a2be2",            # Blue violet
  "Lenarviricota" = "#ff69b4",                 # Hot pink
  "Streptophyta" = "#228b22",                  # Forest green
  "Unclassified viridiplantae" = "#b0c4de",    # Light steel blue
  "Basidiomycota" = "#ff1493",                 # Deep pink
  "Chordata" = "#708090",                      # Slate gray
  "Euglenozoa" = "#8fbc8f",                    # Dark sea green
  "Eustigmatophyte" = "#d8bfd8",               # Thistle
  "Pisuviricota" = "#ee82ee",                  # Violet
  "Chytridiomycota" = "#b22222",                   # Firebrick
  "Nucleocytoviricota" = "#EEC900"
)

                                                 

Ab_NA <- MCycDB_abundance_tax_blast %>%
  # You can uncomment the filter if you want OSPW only
  mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  
  # Step 1: Average CPM per Water_type, Time, and Gene to get the average of the replicates of each gene in each Water_type and Time
   group_by(Water_type, Time, Gene) %>%
  mutate(cpm = mean(CPM), .groups = "drop") %>%


# Step 3: Sum again per Phylum to get the total sum for each Phylum in each Water_type and Time
  group_by(Water_type, Time, Phylum) %>%
  summarise(cpm = sum(cpm)/10^6, .groups = "drop") 

  # Summarize the counts of genes for each Phylum at each level
global_order <- Ab_NA %>%
  group_by(Phylum) %>%
  summarise(sum_cpm = sum(cpm), .groups = "drop") %>%
  # Reorder Phylum based on the number of DEGs within each Sample
  arrange(desc(sum_cpm)) %>%
  mutate(order = row_number())


# Reorder Phylum globally by ensuring unique Phylum levels
Ab_NA1 <- Ab_NA %>%
mutate(Phylum = factor(Phylum, levels = global_order$Phylum))

# Calculate the ordering within each Sample
Ab_NA2 <- Ab_NA1 %>%
  arrange(Time, desc(cpm)) %>%
  mutate(order = row_number())


# Plot the data as a stacked bar plot using cpm
Ab_NA2 %>% 
 mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
 mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  ggplot(aes(x = Time, y = cpm, fill = Phylum, group = order)) +
  geom_bar(stat = "identity", position = "stack") +  # Use 'stack' for stacked bars based on actual percentages
  
  # Add labels and title
  xlab("Time") +
  ylab("Relative abundance (count per million * 10^6)") +
  
  # Customize the y-axis to display exact percentage values
#  scale_y_continuous(limits = c(0, 100)) +  # Set limits from 0 to 100 for percentages
  
  # Customize theme for better readability
  theme(
    text = element_text(size = 16),        # General text size
    axis.title = element_text(size = 16),  # Axis titles font size
    axis.text = element_text(size = 14),   # Axis labels font size
    plot.title = element_text(size = 18),  # Plot title font size
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12),
    
    # Adjust margins and plot layout
    plot.margin = margin(20, 20, 20, 20)
  ) +
  
  # Set legend in two columns for better readability
  guides(fill = guide_legend(ncol = 2)) +
  # Facet the plot by Water_type
  facet_wrap(~ Water_type, ncol = 2) +
  # Apply custom colors using scale_fill_manual
  scale_fill_manual(values = phylum_colors)

ggsave("Figures/Taxa_origins_MCycDB_degraders_Phylum_OSPW_RO.svg", width = 45, height = 35, units = "cm")

Ab_NA %>%
mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
group_by(Water_type, Time) %>%
arrange(.by_group = TRUE, desc(cpm)) %>%
kable(., digits = c(0, 1, 1, 1, 0, 1)) %>%
kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)





# At Order level
# Define custom colors for each Order
order_colors <- c(Burkholderiales = "#dbdb8d", Rhizobiales = "#2ca02c", Caulobacterales = "#ff7f0e", `Unclassified Pseudomonadota` = "#8B6914", Helotiales = "#8c564b", `Unclassified bacteria` = "#7FFF00", Sphingomonadales = "#A52A2A", `Beta Proteobacteria` = "#FF8C00", Flavobacteriales = "#9467bd", Myxococcales = "#EED5B7", Micromonosporales = "#17becf", Peronosporales = "#98df8a", Cytophagales  = "#1f77b4", `Unclassified Bacteroidota` = "#ee82ee", Gemmatimonadales = "#D2691E", Gemmatales = "#FF1493", Pleosporales = "#8B0A50", `Unclassified archaea` = "#aec7e8", Pythiales = "#c49c94", `Alpha Proteobacteria` = "#9edae5", Holophagales = "#d62728", Rhodocyclales = "#bcbd22", Chaetothyriales = "#e377c2", Nostocales = "#0000FF", Sebacinales  = "#458B00", Oomycota = "#EEC900", Leotiomycetes = "#00008B", Saprolegniales = "#8B3A3A", Acidobacteriotales = "#8B6969", Alteromonadales = "#1E90FF", Chytridiales = "#c7c7c7", Nitrosomonadales = "#104E8B", Pseudomonadales = "#FFA500", Chitinophagales = "#CD4F39", `Unclassified Planctomycetota` = "#4682B4", `Unclassified Acidobacteriota` = "#00CD66", `Unclassified Armatimonadota` = "#FF83FA", Bacteriovoracales = "#551A8B", Cellvibrionales = "#008B45", Rhodospirillales = "#2F4F4F", `Gamma Proteobacteria` = "#8B3626", Methanococcales = "#66CDAA")



Ab_MCycDB_order <- MCycDB_abundance_tax_blast %>%
  # You can uncomment the filter if you want OSPW only
  mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  
  # Step 1: Average CPM per Water_type, Time, and Gene to get the average of the replicates of each gene in each Water_type and Time
  group_by(Water_type, Time, Gene) %>%
  mutate(cpm = mean(CPM), .groups = "drop") %>%
  

# Step 3: Sum again per tax_order to get the total sum for each tax_order in each Water_type and Time
  group_by(Water_type, Time, tax_order) %>%
  summarise(cpm = sum(cpm)/10^6, .groups = "drop") 

  # Summarize the counts of genes for each tax_order at each level
global_order <- Ab_MCycDB_order %>%
  group_by(tax_order) %>%
  summarise(sum_cpm = sum(cpm), .groups = "drop") %>%
  # Reorder tax_order based on the number of DEGs within each Sample
  arrange(desc(sum_cpm)) %>%
  mutate(order = row_number())


# Reorder tax_order globally by ensuring unique tax_order levels
Ab_NA1_order <- Ab_MCycDB_order %>%
mutate(tax_order = factor(tax_order, levels = global_order$tax_order))

# Calculate the ordering within each Sample
Ab_NA2_order <- Ab_NA1_order %>%
  arrange(Time, desc(cpm)) %>%
  mutate(order = row_number())


# Plot the data as a stacked bar plot using cpm
Ab_NA2_order %>% 
 mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
 mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
  ggplot(aes(x = Time, y = cpm, fill = tax_order, group = order)) +
  geom_bar(stat = "identity", position = "stack") +  # Use 'stack' for stacked bars based on actual percentages
  
  # Add labels and title
  xlab("Time") +
  ylab("Relative abundance (count per million * 10^3)") +
  
  # Customize the y-axis to display exact percentage values
#  scale_y_continuous(limits = c(0, 100)) +  # Set limits from 0 to 100 for percentages
  
  # Customize theme for better readability
  theme(
    text = element_text(size = 16),        # General text size
    axis.title = element_text(size = 16),  # Axis titles font size
    axis.text = element_text(size = 14),   # Axis labels font size
    plot.title = element_text(size = 18),  # Plot title font size
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12),
    
    # Adjust margins and plot layout
    plot.margin = margin(20, 20, 20, 20)
  ) +
  
  # Set legend in two columns for better readability
  guides(fill = guide_legend(ncol = 2)) +
  # Facet the plot by Water_type
  facet_wrap(~ Water_type, ncol = 2) +
  # Apply custom colors using scale_fill_manual
  scale_fill_manual(values = order_colors)

ggsave("Figures/Taxa_origins_MCycDB_degraders_Order_OSPW_RO.svg", width = 45, height = 35, units = "cm")


Ab_MCycDB_order %>%
mutate(Water_type = factor(Water_type, levels = c("RO", "OSPW"))) %>%
mutate(Time = factor(Time, levels = c("D-0", "D6", "D19", "D40", "D60"))) %>%
group_by(Water_type, Time) %>%
arrange(.by_group = TRUE, desc(cpm)) %>%
kable(., digits = c(0, 1, 1, 1, 0, 1)) %>%
kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```


